app/src/main/java/org/xbmc/kore/ui/sections/remote/ShareHandlingFragment.java

```diff
@@ -0,0 +1,370 @@
+import org.xbmc.kore.jsonrpc.type.PlayerType.GetActivePlayersReturnType;
+import org.xbmc.kore.jsonrpc.type.PlaylistType;
+
+import java.util.ArrayList;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Headless fragment that sends shared URLs to Kodi plugins.
+ */
+public class ShareHandlingFragment extends Fragment {
+
+    /**
+     * Static factory method.
+     *
+     * Creates and attaches a new instance or returns a previously added one.
+     * This should be called early in whatever activity that declares the
+     * intent filters. Sibling fragments can then safely call this and rest
+     * assured that they are getting the same instance.
+     *
+     * @param fm **support** fragment manager.
+     * @return the fragment instance. The host activity should immediately
+     * follow this with a call to {@link #connect(HostConnection)}.
+     */
+    @NonNull
+    public static ShareHandlingFragment of(FragmentManager fm) {
+        Fragment fragment = fm.findFragmentByTag(TAG);
+        if (fragment == null) {
+            fragment = new ShareHandlingFragment();
+            fm.beginTransaction().add(fragment, TAG).commit();
+        }
+        return (ShareHandlingFragment) fragment;
+    }
+
+    private static final String TAG = ShareHandlingFragment.class.getCanonicalName();
+    private static final String KEY_HANDLED = "share.handled";
+    private static final String YOUTUBE_PREFIX = "plugin://plugin.video.youtube/play/?video_id=";
+    private static final String TWITCH_PREFIX = "plugin://plugin.video.twitch/playLive/";
+    private static final String VIMEO_PREFIX = "plugin://plugin.video.vimeo/play/?video_id=";
+    private static final String YOUTUBE_SHORT_URL = "://youtu\\.be/([^\\?\\s/]+)";
+    private static final String YOUTUBE_LONG_URL = "://(?:www\\.)?youtube\\.com/watch\\S*[\\?&]v=([^&\\s]+)";
+    private static final String TWITCH_URL = "://www\\.twitch\\.tv/([^\\?\\s/]+)";
+    private static final String VIMEO_URL = "://(?:www\\.|player\\.)?vimeo\\.com[^\\?\\s]*?/(\\d+)";
+
+    interface ShareHandler {
+        /**
+         * @return null when no url is recognized from this piece of data
+         * passed through the intent.
+         */
+        @Nullable String urlFrom(@Nullable String source);
+    }
+
+    interface OnFinish<T> {
+        void got(T result);
+    }
+
+    interface Task<T> {
+        void start(@NonNull OnFinish<? super T> then);
+    }
+
+    static class Seq<T> implements Task<T> {
+        private final Task<T> task;
+
+        Seq(Task<T> task) {
+            this.task = task;
+        }
+
+        <U> Seq<U> then(final Task<U> block) {
+            return new Seq<>(new Task<U>() {
+                @Override
+                public void start(@NonNull final OnFinish<? super U> then) {
+                    task.start(new OnFinish<T>() {
+                        @Override
+                        public void got(T unused) {
```

Comment:
That's how it works currently. The downstream tasks aren't executed when a task doesn't invoke its `OnFinish`. I'll admit the code is really confusing as short as it is, but in my log tests before I removed them, it appears correct. I'll try to write tests to confirm.